### ISSUE TITLE
error is produced with autoreload

### ISSUE THREAD

**Issue by ceilingcat** [ipython/ipython#9349]
I found an issue with autoreload generating unexpected errors. Originally I asked this question on stackoverflow (http://stackoverflow.com/questions/36174311/how-does-autoreload-work-with-jupyter) because I thought this might be a silly mistake on my part, but according to takluyver this might be a bug, so I am filing a report here.

I have:
Anaconda 2.5.0, which installed Python 3.5.0, IPython 4.1.2, jupyter 4.1.0
Mac OS X 10.11.2

In this MWE, I have a file `mytest.py` in the working directory, with the following content:

    class MyTest(object):
    
      def __init__(self):
        print(1)
        pass

In `jupyter` I have these cells.

Cell 1:

    %reload_ext autoreload
    %autoreload 2
    import mytest

Cell 2:

    x = mytest.MyTest()

If I start with a fresh kernel then it works as expected, and it gives me `1`. Now if I go into `mytest.py` and change the print out to 2 and rerun Cell 2 I would get:

    1
    [autoreload of mytest failed: Traceback (most recent call last):
      File "/Users/myusername/anaconda/lib/python3.5/site-packages/IPython/extensions/autoreload.py", line 247, in check
        superreload(m, reload, self.old_objects)
    NameError: name 'reload' is not defined
    ]
    [autoreload of mytest failed: Traceback (most recent call last):
      File "/Users/myusername/anaconda/lib/python3.5/site-packages/IPython/extensions/autoreload.py", line 247, in check
        superreload(m, reload, self.old_objects)
    NameError: name 'reload' is not defined
    ]

If I run Cell 2 again now I'd get just `1` without any errors, but it still has not used the new code.

I can't figure out how to reload the module properly without restarting the kernel, which I can't do because I have some very time-consuming code in the real program.

I have tried this in `IPython` in terminal. Although instantiating the class after modifying the code also gives me errors (note that it's a different error: `AttributeError: 'NoneType' object has no attribute '__dict__'`) but it would use the updated code if I run it again. So it seems to me that this is a problem specific to `jupyter`.

This seems to be different than https://github.com/ipython/ipython/issues/8347 because I can consistently reproduce it. It also applies to existing methods too.

**Comment by Carreau** [comment]
hum.. that seem weird, I have issues to reproduce this locally:

```
$ ipython
Python 3.5.1 |Continuum Analytics, Inc.| (default, Dec  7 2015, 11:24:55)
Type "copyright", "credits" or "license" for more information.

IPython 5.0.0.dev -- An enhanced Interactive Python.
?         -> Introduction and overview of IPython's features.
%quickref -> Quick reference.
help      -> Python's own help system.
object?   -> Details about 'object', use 'object??' for extra details.

...

In [3]: !mvim mytest.py

In [4]: %reload_ext autoreload
   ...: %autoreload 2
   ...: import mytest

In [5]: x = mytest.MyTest()
1

In [6]: x = mytest.MyTest()
2

In [7]: !cat mytest.py
class MyTest(object):

  def __init__(self):
    print(2)
    pass
```

**Comment by ceilingcat** [comment]
sorry for the newbie question... do you know what is the most painless way to install (and later uninstall) IPython 5.0.0.dev with anaconda?

**Comment by Carreau** [comment]
Does not seem to be a 4.1 thing either, as I cannot reproduce on it either. 

But to answer your question:

```
git clone https://github.com/ipython/ipython
cd ipython
pip install .
```

(replace `pip install .` by `pip install -e .` if you want to have a development install and want to be able to change the source) 

See [there](http://ipython.readthedocs.org/en/latest/install/install.html#installing-ipython-itself) for more info.

**Comment by ceilingcat** [comment]
I found that the error is caused by having

    c.InteractiveShell.deep_reload = True

in `ipython_config.py`. Commenting it out fixed the problem in both IPython (in Terminal) and jupyter.

Sorry for not looking in my config file first.

**Comment by takluyver** [comment]
No problem. Yes, I think deep reload and autoreload are not likely to work together. You shouldn't need both, in any case.

**Comment by absognety** [comment]
This issue occurred with `Python 3.9.0` and `IPython 7.21.0`, but resolved when I bumped the `IPython` to `7.23.0`
